### PR TITLE
fix: OOO for team events

### LINKED ISSUE #14507: [CAL-3391] Fix OOO for team events
Team events currently don't handle OOO correctly.

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-3391](https://linear.app/calcom/issue/CAL-3391/fix-ooo-for-team-events)</sub>

### PR DESCRIPTION
**Pull request by CarinaWolli**
## What does this PR do?

Fixes #14507

This fixes availabilities for team events when users have OOO entries. With this PR we now already subtract the OOO days from the user's date ranges and there are no OOO emojis or redirection buttons shown.

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How should this be tested?

- test availabilities of round robin and collective events when users have OOO entries

### CODE REVIEW

File: packages/lib/date-ranges.ts [hunk 1]
```diff
@@ -158,7 +178,16 @@ export function buildDateRanges({
...
   );
 
-  return dateRanges.flat();
+  const dateRangesWithoutOOO = Object.values({
+    ...groupedWorkingHours,
+    ...groupedDateOverrides,
+    ...groupedOOO,
+  }).map(
+    // remove 0-length overrides && OOO dates that were kept to cancel out working dates until now.
+    (ranges) => ranges.filter((range) => range.start.valueOf() !== range.end.valueOf())
```
Comment (CarinaWolli) [inline packages/lib/date-ranges.ts:187]: OOO always starts and end on the same time (handled the same way as a full-day date override)

File: packages/core/getAggregatedAvailability/index.ts [hunk 2]
```diff
@@ -5,18 +5,28 @@ import { SchedulingType } from "@calcom/prisma/enums";
...
 export const getAggregatedAvailability = (
-  userAvailability: { dateRanges: DateRange[]; user?: { isFixed?: boolean } }[],
+  userAvailability: {
+    dateRanges: DateRange[];
+    dateRangesWithoutOOO: DateRange[];
+    user?: { isFixed?: boolean };
+  }[],
   schedulingType: SchedulingType | null
 ): DateRange[] => {
+  const isTeamEvent =
```
Comment (CarinaWolli) [inline packages/core/getAggregatedAvailability/index.ts:15]: for team events we want to use the date ranges that already take into account OOO days

File: packages/trpc/server/routers/viewer/slots/util.ts [hunk 3]
```diff
@@ -532,7 +539,7 @@ export async function getAvailableSlots({ input, ctx }: GetScheduleOptions): Pro
     frequency: eventType.slotInterval || input.duration || eventType.length,
     organizerTimeZone:
       eventType.timeZone || eventType?.schedule?.timeZone || allUsersAvailability?.[0]?.timeZone,
-    datesOutOfOffice: allUsersAvailability[0]?.datesOutOfOffice,
+    datesOutOfOffice: !isTeamEvent ? allUsersAvailability[0]?.datesOutOfOffice : undefined,
```
Comment (CarinaWolli) [inline packages/trpc/server/routers/viewer/slots/util.ts:542]: For team events, OOO days are already included `aggregatedAvailability`

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 4]
```diff
@@ -357,7 +357,7 @@ export async function ensureAvailableUsers(
   }
 
   for (const user of eventType.users) {
-    const { dateRanges, busy: bufferedBusyTimes } = await getUserAvailability(
+    const { oooExcludedDateRanges: dateRanges, busy: bufferedBusyTimes } = await getUserAvailability(
```
Comment (CarinaWolli) [inline packages/features/bookings/lib/handleNewBooking.ts:360]: we never want a user to be booked when marked as OOO

**Review by zomars — APPROVED** [review]
LGTM

**Pull request merged**